api.md

```diff
@@ -390,6 +391,242 @@ if you try to upload another file other than an image:
+    conn = post(conn, ~p"/api/images", @empty_image)
+
+    assert Map.get(Jason.decode!(response(conn, 400)), "errors") == %{
+             "detail" => "Error uploading file. Failure creating the CID filename."
+           }
+  end
+
+  test "valid file but the upload to S3 failed. It should return an error.", %{conn: conn} do
+
+    with_mock ExAws, [request: fn(_input) -> {:error, :failure} end] do
+      conn = post(conn, ~p"/api/images", @valid_image_attrs)
+
+      assert Map.get(Jason.decode!(response(conn, 400)), "errors") == %{
+               "detail" => "Error uploading file. There was an error uploading the file to S3."
+             }
+    end
+  end
+```
+
+> **Note** 
+>
+> We are using the [`mock` ](https://github.com/jjh42/mock)
+> library so we are able to correctly test 
+> whenever `S3` is down.
+> Although using mocks *can lead to false positives*,
+> here's acceptable because we're simulating a third-party service to be down.
+
+
+We've added a test for each scenario 
+and what we expect the API to return to us.
+We are using the `empty` file
+to simulate an invalid binary file content,
+which will result in failing the `CID` creation.
+Check the files needed for these test to pass inside
+[`priv/static/images`](./priv/static/images/).
+
+Now let's implement the features so our tests pass! ✅
+
+Open `lib/app/upload.ex`.
+We are going to refactor this function
+so it returns specific errors 
+as the image is read all the way until it is uploaded to `S3`.
+
+```elixir
+def upload(image) do
+    # Create `CID` from file contents so filenames are unique
+    case File.read(image.path) do
+      {:ok, file_binary} ->
+        contents = if byte_size(file_binary) == 0, do: [], else: file_binary
+        file_cid = Cid.cid(contents)
+
+        file_extension =
+          image.content_type
+          |> MIME.extensions()
+          |> List.first()
+
+        # Check if file `cid` and extension are valid.
+        case {file_cid, file_extension} do
+          {"invalid data type", nil} ->
+            {:error, :invalid_extension_and_cid}
+
+          {"invalid data type", _extension} ->
+            {:error, :invalid_cid}
+
+          {_cid, nil} ->
+            {:error, :invalid_extension}
+
+          # If the `cid` and extension are valid, we are safe to upload
+          {file_cid, file_extension} ->
+            # Creating filename with the retrieved extension
+            file_name = "#{file_cid}.#{file_extension}"
+
+            # Upload to S3
+            request_response =
+              try do
+                image.path
+                |> ExAws.S3.Upload.stream_file()
+                |> ExAws.S3.upload(Application.get_env(:ex_aws, :original_bucket), file_name,
+                  acl: :public_read,
+                  content_type: image.content_type
+                )
+                |> ExAws.request(get_ex_aws_request_config_override())
+              rescue
+                _e ->
```

Comment:
just added a `Logger` line to the `upload_file_to_s3/3` function in `lib/app/upload.ex`:
https://github.com/dwyl/imgup/blob/831d2eb4cec9c20ed09b822e38a579d8ef22f6d6/lib/app/upload.ex#L79

Guess what it logs as part of the `mock` invocation ...
https://github.com/dwyl/imgup/actions/runs/5560741019/jobs/10157932627#step:7:216
<img width="1487" alt="image" src="https://github.com/dwyl/imgup/assets/194400/0c5b60b2-6d56-4230-8094-e39e40de8fc5">

```sh
Error: 00:35.176 request_id=F3H1G9fWV3C_m1oAAbzS [error] ** (UndefinedFunctionError) function ExAws.request/2 is undefined or private
    (ex_aws 2.4.3) ExAws.request(%ExAws.S3.Upload{src: %File.Stream{path: "/home/runner/work/imgup/imgup/_build/test/lib/app/priv/static/images/phoenix.png", modes: [:raw, :read_ahead, :binary], line_or_bytes: 5242880, raw: true}, bucket: "imgup-original", path: "zb2rhXACvyoVCaV1GF5ozeoNCXYdxcKAEWvBTpsnabo3moYwB.png", upload_id: nil, opts: [acl: :public_read, content_type: "image/png"], service: :s3}, %{})
    (app 1.0.0) lib/app/upload.ex:74: App.Upload.upload_file_to_s3/3
    (app 1.0.0) lib/app/upload.ex:26: App.Upload.upload/1
```

`UndefinedFunctionError`. 
So not only was the function not logging an error but the `mock` is a **`false` positive**. 
This is not "testing" an error on `S3`; it's _mocking_ a function call incorrectly. 

![picard-double-facepalm](https://github.com/dwyl/imgup/assets/194400/387e84a1-c79f-4761-9a76-fc00903e9353)

Reply:
I've removed this `mock` because it doesn't add any robustness to the project. 
I've left your code 99.9% intact and only added the `Logger.error` in case there is an `S3` upload error so that we can trace and resolve it.